### PR TITLE
Create base CRUD

### PR DESCRIPTION
**Pull request by ferdn4ndo**

### CODE REVIEW

File: src/app/clients/client_base.py [hunk 1]
```diff
@@ -1,7 +1,14 @@
+import functools
 from abc import ABC, abstractmethod
+from collections import Callable
+
+from tornado.ioloop import IOLoop
```
Comment (adrianoadum) [inline src/app/clients/client_base.py]: unused imports
```suggestion
from abc import ABC, abstractmethod
```

File: src/app/commands/fetch_candlestick_history.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
+import logging
+
+from app.clients.binance.binance_client import BinanceClient
```
Comment (adrianoadum) [inline src/app/commands/fetch_candlestick_history.py]: unused import
```suggestion
```

File: src/app/commands/fetch_candlestick_history.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+from app.errors import ResourceNotFoundError, ResourceAlreadyInUseError
+from app.services import DatabaseService
+from app.services.consumer_service import ConsumerService
+from app.services.consumer_service_factory import create_client_from_exchange_code, create_service_from_exchange_code
+from app.services.currency_pair.currency_pair_service import CurrencyPairService
+from app.services.exchanges.binance_exchange_service import BinanceExchangeService
+from sqlalchemy.orm import Session
```
Comment (adrianoadum) [inline src/app/commands/fetch_candlestick_history.py]: unused imports
```suggestion
```

File: src/app/commands/init_binance_websocket.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    Usage:
+    python manage.py init_binance_websocket PAIR1 PAIR2 ...
+"""
+
+
+def execute(arguments: list) -> None:
+    pairs = arguments[1:]
+
+    logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s", datefmt="%Y-%m-%d %H:%M:%S")
+
```
Comment (adrianoadum) [inline src/app/commands/init_binance_websocket.py:24]: Is it necessary to set the logger config here? Isn't the config made in the src/app/\_\_init\_\_.py enough?

File: src/app/commands/manage_binance_websocket_service.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+
+def execute(arguments: list) -> None:
+    operation = arguments[1]
+    pairs = arguments[2:]
+
+    if operation == "add_pair":
+        binance_websocket_service.register_pair(pairs)
+    else:
+        print(f"The operation '${operation}' is not supported!")
```
Comment (adrianoadum) [inline src/app/commands/manage_binance_websocket_service.py]: Maybe I'm wrong, but the $ isn't necessary
```suggestion
        print(f"The operation '{operation}' is not supported!")
```

File: src/app/commands/populate_exchange_data.py [hunk 6]
```diff
@@ -1,48 +1,30 @@
+import logging
+
 from app.clients.binance.binance_client import BinanceClient
 from app.clients.client_exception import ClientException
```
Comment (adrianoadum) [inline src/app/commands/populate_exchange_data.py:4]: unused imports
```suggestion
from app.clients.binance.binance_client import BinanceClient
from app.clients.client_exception import ClientException
```

File: src/app/controllers/exchange_historical_controller.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
+from datetime import datetime
+import json
+import logging
+
+from tornado import gen
+from tornado.ioloop import IOLoop
```
Comment (adrianoadum) [inline src/app/controllers/exchange_historical_controller.py]: unused import
```suggestion
```

File: src/app/controllers/exchange_historical_controller.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+from tornado import concurrent
+
+from app.controllers.base_controller import BaseController
+from app.errors import ResourceAlreadyInUseError
+from app.services.consumer_service import ConsumerService
+from app.services.consumer_service_factory import (
+    create_consumer_service_from_exchange,
```
Comment (adrianoadum) [inline src/app/controllers/exchange_historical_controller.py]: unused import
```suggestion
```

File: src/app/controllers/exchange_historical_controller.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    create_client_from_exchange_code,
+    create_service_from_exchange_code,
+)
+from app.services.currency_pair.currency_pair_service import CurrencyPairService
+from app.services.database_service import DatabaseService
+from app.services.exchanges.exchange_service_factory import create_exchange_service_from_id, get_exchange_by_id
```
Comment (adrianoadum) [inline src/app/controllers/exchange_historical_controller.py]: unused import
```suggestion
from app.services.exchanges.exchange_service_factory import get_exchange_by_id
```

File: src/app/controllers/exchange_setup_controller.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
+import logging
+
+from app.clients.client_exception import ClientException
+from app.controllers.base_controller import BaseController
+from app.errors.model_already_exists_error import ModelAlreadyExistsError
+from app.errors.resource_not_found_error import ResourceNotFoundError
+from app.models import Exchange
+from app.services.consumer_service_factory import create_client_from_exchange_code, create_service_from_exchange_code
```
Comment (adrianoadum) [inline src/app/controllers/exchange_setup_controller.py]: unused imports
```suggestion
```

File: src/app/controllers/exchange_setup_controller.py [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+from app.services.database_service import DatabaseService
+from app.services.exchanges.exchange_service_factory import (
+    ALLOWED_EXCHANGE_CODES,
+    create_exchange_service_from_code,
+    get_exchange_by_id,
+    serialize_existing_exchanges,
+    create_exchange_service_from_model,
+)
```
Comment (adrianoadum) [inline src/app/controllers/exchange_setup_controller.py]: unused imports
```suggestion
from app.services.exchanges.exchange_service_factory import get_exchange_by_id
```

File: src/app/services/consumer_service.py [hunk 12]
```diff
@@ -1,47 +1,75 @@
-from datetime import datetime
+import asyncio
+import datetime
+import functools
```
Comment (adrianoadum) [inline src/app/services/consumer_service.py]: unused imports
```suggestion
import datetime
```

File: src/app/services/consumer_service.py [hunk 13]
```diff
@@ -1,47 +1,75 @@
...
+import logging
+
+from sqlalchemy.orm import Session
+from tornado import gen
 
 from app.clients.client_base import ClientBase
 from app.clients.client_exception import ClientException
+from app.errors import ResourceAlreadyInUseError, ClientIntegrationError
```
Comment (adrianoadum) [inline src/app/services/consumer_service.py]: unused import
```suggestion
from app.errors import ClientIntegrationError
```

File: src/app/services/consumer_service.py [hunk 14]
```diff
@@ -1,47 +1,75 @@
...
 from app.models import CurrencyPair, Exchange
+from app.services.currency_pair.currency_pair_factory import get_exchange_currency_pair_from_symbol
```
Comment (adrianoadum) [inline src/app/services/consumer_service.py]: unused import
```suggestion
```

File: src/app/services/consumer_service.py [hunk 15]
```diff
@@ -1,47 +1,75 @@
...
+from app.services.currency_pair.currency_pair_service import CurrencyPairService
+from app.services.database_service import DatabaseService
 from app.services.exchanges.exchange_service_base import ExchangeServiceBase
-from sqlalchemy.orm.exc import NoResultFound
+from app.services.exchanges.exchange_service_factory import create_exchange_service_from_code
```
Comment (adrianoadum) [inline src/app/services/consumer_service.py]: unused import
```suggestion
```

File: src/app/services/currency_pair/currency_pair_service.py [hunk 16]
```diff
@@ -0,0 +1,86 @@
+import logging
+
+from datetime import datetime
```
Comment (adrianoadum) [inline src/app/services/currency_pair/currency_pair_service.py]: unused import
```suggestion
```

File: src/app/services/database_service.py [hunk 17]
```diff
@@ -1,33 +1,43 @@
 import os
```
Comment (adrianoadum) [inline src/app/services/database_service.py]: unused import
```suggestion
```

File: src/app/services/database_service.py [hunk 18]
```diff
@@ -58,6 +93,7 @@ def _create_object_from_params(self, model, lookup: dict, params: dict, lock=Fal
             with self.session.begin_nested():
                 self.session.flush()
         except IntegrityError:
+            print("maoe")
```
Comment (adrianoadum) [inline src/app/services/database_service.py]:
```suggestion
```

File: src/app/services/exchanges/exchange_service_base.py [hunk 19]
```diff
@@ -1,43 +1,67 @@
...
+    def __init__(self, exchange: Exchange = None) -> None:
+        self.exchange = exchange
 
     def add_exchange(self) -> Exchange:
-        (exchange, _) = self.database.update_or_create(
-            Exchange,
-            code=self.EXCHANGE_CODE,
-            defaults={"name": self.EXCHANGE_NAME},
-        )
+
```
Comment (adrianoadum) [inline src/app/services/exchanges/exchange_service_base.py]:
```suggestion
```

File: src/app/services/setup_service.py [hunk 20]
```diff
@@ -0,0 +1,27 @@
+import logging
+
+from sqlalchemy.orm import Session
+
+from app.services.consumer_service_factory import create_client_from_exchange_code, create_service_from_exchange_code
+from app.services.database_service import DatabaseService
```
Comment (adrianoadum) [inline src/app/services/setup_service.py]: unused import
```suggestion
```

File: src/app/services/database_service.py (hunk 18 above)
Comment (ferdn4ndo) [reply src/app/services/database_service.py]: maoe

**Pull request merged**